### PR TITLE
adds support for uart 4-wire mode operation and parity generation/checking 

### PR DESCRIPTION
**Pull request by kritikbhimani**

### CODE REVIEW

File: src/main/scala/devices/uart/UART.scala [hunk 1]
```diff
@@ -18,16 +19,21 @@ case class UARTParams(
   oversample: Int = 4,
   nSamples: Int = 3,
   nTxEntries: Int = 8,
-  nRxEntries: Int = 8)
+  nRxEntries: Int = 8,
+  wire4: Boolean = false,
```
Comment (mwachs5) [inline src/main/scala/devices/uart/UART.scala]: Can these names be more descriptive/less abbreviated? I might suggest:
-includeFourWire
-includeParity
-include???Parity (I'm not sure what `ind` is)

File: src/main/scala/devices/uart/UART.scala [hunk 2]
```diff
@@ -127,6 +150,24 @@ abstract class UART(busWidthBytes: Int, val c: UARTParams, divisorInit: Int = 0)
       RegField(c.divisorBits, div,
                  RegFieldDesc("div","Baud rate divisor",reset=Some(divisorInit))))
   )
+
+  val optionalparity = if (c.parity) Seq(
+    UARTCtrlRegs.parity -> RegFieldGroup("parity",Some("Odd/Even Parity Generation/Checking"),Seq(
```
Comment (mwachs5) [inline src/main/scala/devices/uart/UART.scala]: can the register group have a different name than the subfields? it is potentially confusing to use `parity` for both.

File: src/main/scala/devices/uart/UARTRx.scala [hunk 3]
```diff
@@ -3,16 +3,22 @@ package sifive.blocks.devices.uart
...
-import freechips.rocketchip.util.Majority
+import freechips.rocketchip.util._
 
 class UARTRx(c: UARTParams) extends Module {
   val io = new Bundle {
     val en = Bool(INPUT)
     val in = Bits(INPUT, 1)
     val out = Valid(Bits(width = c.dataBits))
     val div = UInt(INPUT, c.divisorBits)
+    val enparity = if (c.parity) Some(Bool(INPUT)) else None
```
Comment (mwachs5) [inline src/main/scala/devices/uart/UARTRx.scala]: you should use the Option utility: shortens to `c.parity.option(Bool(INPUT)`

File: src/main/scala/devices/uart/UART.scala [hunk 4]
```diff
@@ -18,16 +18,19 @@ case class UARTParams(
   oversample: Int = 4,
   nSamples: Int = 3,
   nTxEntries: Int = 8,
-  nRxEntries: Int = 8)
+  nRxEntries: Int = 8,
+  wire4: Boolean = false)
 {
```
Comment (KowshikSiFive) [inline src/main/scala/devices/uart/UART.scala]: Could you also push your real-time configurable 8/9 bit data frame changes here?

File: src/main/scala/devices/uart/UARTRx.scala [hunk 5]
```diff
@@ -76,11 +84,23 @@ class UARTRx(c: UARTParams) extends Module {
         sample_count := countdown(c.oversample-1, 0)
 
         when (sample_mid) {
-          when (data_last) {
-            state := s_idle
-            valid := Bool(true)
-          } .otherwise {
-            shifter := Cat(voter, shifter >> 1)
+          if (c.includeParity) {
```
Comment (mhtwn) [inline src/main/scala/devices/uart/UARTRx.scala]: Some comments about what logic is doing would be useful for maintainability going forward!

File: src/main/scala/devices/uart/UARTTx.scala [hunk 6]
```diff
@@ -27,16 +32,29 @@ class UARTTx(c: UARTParams) extends Module {
...
-    counter := Mux1H((0 until c.stopBits).map(i =>
-      (io.nstop === UInt(i)) -> UInt(n + i + 1)))
+    if (c.includeParity) {
+      val includebit9 = if (c.dataBits == 9) Mux(io.data8or9.get, Bool(false), io.in.bits(8)) else Bool(false)
+      val parity = Mux(io.enparity.get, includebit9 ^ io.in.bits(7,0).toBools.reduce(_ ^ _) ^ io.parity.get, Bool(true))
+      val paritywithbit9 = if (c.dataBits == 9) Mux(io.data8or9.get, Cat(1.U(1.W), parity), Cat(parity, io.in.bits(8))) 
+                           else Cat(1.U(1.W), parity)
+      shifter := Cat(paritywithbit9, io.in.bits(7,0), Bits(0, 1))
+      counter := Mux1H((0 until c.stopBits).map(i =>
+        (io.nstop === UInt(i)) -> UInt(n + i + 1))) - (!io.enparity.get).asUInt - io.data8or9.getOrElse(0.U)
```
Comment (mhtwn) [inline src/main/scala/devices/uart/UARTTx.scala]: Again, I would highly encourage adding comments here too

**Comment by kritikbhimani** [comment]
reopened here #141

**Pull request closed without merging**